libnavigation-core/src/main/java/com/mapbox/navigation/core/history/MapboxHistoryRecorder.kt

```diff
@@ -0,0 +1,41 @@
+package com.mapbox.navigation.core.history
+
+import com.mapbox.navigation.base.options.HistoryRecorderOptions
+import com.mapbox.navigation.core.MapboxNavigation
+
+/**
+ * Interface into retrieving mapbox navigation history files. Retrieve an instance
+ * of this interface through [MapboxNavigation.historyRecorder].
+ *
+ * @see [HistoryRecorderOptions] to enable the recorder.
+ */
+interface MapboxHistoryRecorder {
```

Comment:
```suggestion
sealed interface MapboxHistoryRecorder {
```
We don't actually want custom implementations of this interface.

Also cc @Guardiola31337, we found a use case!

Reply:
can't have sealed interfaces.. there's no real use to overriding your own MapboxHistoryRecorder, as there is no way to set it.

Otherwise we start thinking about making many of the interfaces sealed classes. I'm not seeing the reason for this though

Reply:
> can't have sealed interfaces.. there's no real use to overriding your own MapboxHistoryRecorder, as there is no way to set it.

Isn't that the reason to have a `sealed interface`?

> Otherwise we start thinking about making many of the interfaces sealed classes. I'm not seeing the reason for this though

What's the problem of `start thinking about making many of the interfaces sealed classes`? 🤔

Reply:
I mean `sealed interface` will not compile 😄 

![Screen Shot 2021-06-11 at 9 03 38 AM](https://user-images.githubusercontent.com/3021882/121715648-fb4c0a80-ca93-11eb-8e8d-c1d8ef355da9.png)

Reply:
https://github.com/mapbox/mapbox-navigation-android/pull/2689#discussion_r402557516

I did try it though. Also we had this conversation before 😆

Reply:
That's an IDE issue, I also see that highlighted in red but it compiles and works fine at the end. Refs. https://github.com/mapbox/mapbox-navigation-android/issues/4374

Reply:
This is a perfect reason to expose a sealed interface because _developers should not implement it_. They only invoke functions of that interface. By making it sealed, we can keep adding a method to this interface without introducing breaking changes. If not a sealed interface, we should simply expose a class since we're not planning to have other implementations of the HistoryRecorder. I understand it goes against the principle of "programming against interfaces", but in our environment, this is a double-edged sword due to semantic versioning.

Reply:
In order to see what is going on here, you need to look at the `HistoryRecorderController`. It is the only implementer of `MapboxHistoryRecorder`, and it exposes a few functions so it can change the `HistoryRecorderHandle` under the hood. 

> e: /Users/kylemadsen/Development/mapbox/mapbox-navigation-android/libnavigation-core/src/main/java/com/mapbox/navigation/core/history/HistoryRecorderController.kt: (30, 36): This type is sealed, so it can be inherited by only its own nested classes or objects

> If not a sealed interface, we should simply expose a class since we're not planning to have other implementations of the HistoryRecorder.

If we do this, we require downstream users to solve lifecycle issues for semantics. The approach here is optimizing for simpler integrations.

Reply:
We could expose `HistoryRecorderController` and rename it `MapboxHistoryRecorder` 

Remove the interface